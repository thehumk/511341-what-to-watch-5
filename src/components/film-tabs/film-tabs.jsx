import React from 'react';
import PropTypes from 'prop-types';
import {propsForFilms, propsForComments} from '../../utils/prop-types';
import {getFilmRating} from '../../utils/film';
import {setFormatCommentDate} from '../../utils/utils';
import {withFilmTabs} from '../../hocs/with-film-tabs/with-film-tabs';

const TypeTabs = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};

const FilmTabs = (props) => {
  const {film, filmComments, tab, tabsClickHandler} = props;
  const textRating = getFilmRating(film.rating);
  const duration = (film.run_time / 60 | 0) + `h ` + (film.run_time % 60) + `m`;

  return (
    <div className="movie-card__desc">
      <nav className="movie-nav movie-card__nav">
        <ul className="movie-nav__list">
          <li className={`movie-nav__item ${tab === TypeTabs.OVERVIEW ? `movie-nav__item--active` : ``}`}>
            <a href="#" className="movie-nav__link" onClick={(evt) => {
              tabsClickHandler(evt, TypeTabs.OVERVIEW);
            }}>Overview</a>
          </li>
          <li className={`movie-nav__item ${tab === TypeTabs.DETAILS ? `movie-nav__item--active` : ``}`}>
            <a href="#" className="movie-nav__link" onClick={(evt) => {
              tabsClickHandler(evt, TypeTabs.DETAILS);
            }}>Details</a>
          </li>
          <li className={`movie-nav__item ${tab === TypeTabs.REVIEWS ? `movie-nav__item--active` : ``}`}>
            <a href="#" className="movie-nav__link" onClick={(evt) => {
              tabsClickHandler(evt, TypeTabs.REVIEWS);
            }}>Reviews</a>
          </li>
          {}
        </ul>
      </nav>

      {tab === TypeTabs.OVERVIEW && (
        <>
          <div className="movie-rating">
            <div className="movie-rating__score">{film.rating}</div>
            <p className="movie-rating__meta">
              <span className="movie-rating__level">{textRating}</span>
              <span className="movie-rating__count">{film.scores_count} ratings</span>
            </p>
          </div>

          <div className="movie-card__text">
            <p>{film.description}</p>

            <p className="movie-card__director"><strong>Director: {film.director}</strong></p>

            <p className="movie-card__starring"><strong>Starring: {film.starring.join(`, `)} and other</strong></p>
          </div>
        </>
      )}

      {tab === TypeTabs.DETAILS && (
        <div className="movie-card__text movie-card__row">
          <div className="movie-card__text-col">
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Director</strong>
              <span className="movie-card__details-value">{film.director}</span>
            </p>
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Starring</strong>
              <span className="movie-card__details-value">
                {film.starring.map((elem, i) => (
                  <React.Fragment key={i}>
                    {elem}
                    {i < film.starring.length - 1 ? `, ` : ``}
                    <br/>
                  </React.Fragment>
                ))}
              </span>
            </p>
          </div>

          <div className="movie-card__text-col">
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Run Time</strong>
              <span className="movie-card__details-value">{duration}</span>
            </p>
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Genre</strong>
              <span className="movie-card__details-value">{film.genre}</span>
            </p>
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Released</strong>
              <span className="movie-card__details-value">{film.released}</span>
            </p>
          </div>
        </div>
      )}

      {tab === TypeTabs.REVIEWS && (
        <div className="movie-card__reviews movie-card__row">
          <div className="movie-card__reviews-col">
            {filmComments.map((elem, i) => (
              <div className="review" key={i}>
                <blockquote className="review__quote">
                  <p className="review__text">{elem.comment}</p>

                  <footer className="review__details">
                    <cite className="review__author">{elem.user.name}</cite>
                    <time className="review__date" dateTime="2016-12-24">{setFormatCommentDate(elem.date)}</time>
                  </footer>
                </blockquote>

                <div className="review__rating">{elem.rating}</div>
              </div>
            ))}
          </div>
        </div>
      )}
    </div>
  );
};

FilmTabs.propTypes = {
  film: propsForFilms,
  filmComments: PropTypes.arrayOf(propsForComments).isRequired,
  tab: PropTypes.string.isRequired,
  tabsClickHandler: PropTypes.func.isRequired,
};

export {FilmTabs};
export default withFilmTabs(FilmTabs);
