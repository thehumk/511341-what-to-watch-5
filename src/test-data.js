export const noop = () => {};

export const MATCH = {params: {id: `1`}};

export const mockComments = [
  {
    comment: `123`,
    date: `2020-10-29T13:38:44.769Z`,
    id: 1,
    rating: 4.6,
    user: {
      id: 12,
      name: `Isaac`
    },
  },
  {
    comment: `1234`,
    date: `2020-10-29T13:38:45.769Z`,
    id: 2,
    rating: 4.5,
    user: {
      id: 13,
      name: `Isa`
    },
  }
];

/* eslint-disable */

export const mockFilms = [
  {
    background_color: `#73B39A`,
    background_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    director: `Nicolas Winding Refn`,
    genre: `Action`,
    id: 1,
    is_favorite: true,
    name: `Bronson`,
    poster_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 3.6,
    released: 2008,
    run_time: 92,
    scores_count: 109661,
    starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    video_link: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  },
  {
    background_color: `#73B39A`,
    background_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    director: `Nicolas Winding Refn`,
    genre: `Drama`,
    id: 2,
    is_favorite: false,
    name: `Bronson`,
    poster_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 3.6,
    released: 2008,
    run_time: 92,
    scores_count: 109661,
    starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    video_link: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  },
];

/* eslint-enable */

export const state = {
  DATA: {
    films: mockFilms,
    promoFilm: mockFilms[0],
    genresList: [`All`, `Action`, `Drama`],
    isApplicationReady: false,
  },
  FILMS_STATUS: {
    activeGenre: `All`,
    renderFilmsCount: 8,
  },
  USER: {
    authorizationStatus: `NO_AUTH`,
    user: null,
  },
  COMMENTS: {
    comments: mockComments,
  },
};
