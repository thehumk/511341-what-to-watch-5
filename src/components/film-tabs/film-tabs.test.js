import React from 'react';
import renderer from 'react-test-renderer';
import {FilmTabs} from './film-tabs';
import {noop, mockFilms, mockComments} from '../../test-data';

it(`Should FilmTabs render correctly`, () => {
  const tree = renderer.create(
      <FilmTabs
        film={mockFilms[0]}
        filmComments={mockComments}
        tab={`Reviews`}
        tabsClickHandler={noop}
      />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
