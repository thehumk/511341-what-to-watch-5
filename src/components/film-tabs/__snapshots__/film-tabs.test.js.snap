// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should FilmTabs render correctly 1`] = `
<div
  className="movie-card__desc"
>
  <nav
    className="movie-nav movie-card__nav"
  >
    <ul
      className="movie-nav__list"
    >
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Overview
        </a>
      </li>
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Details
        </a>
      </li>
      <li
        className="movie-nav__item movie-nav__item--active"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Reviews
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="movie-card__reviews movie-card__row"
  >
    <div
      className="movie-card__reviews-col"
    >
      <div
        className="review"
      >
        <blockquote
          className="review__quote"
        >
          <p
            className="review__text"
          >
            123
          </p>
          <footer
            className="review__details"
          >
            <cite
              className="review__author"
            >
              Isaac
            </cite>
            <time
              className="review__date"
              dateTime="2016-12-24"
            >
              October 29, 2020
            </time>
          </footer>
        </blockquote>
        <div
          className="review__rating"
        >
          4.6
        </div>
      </div>
      <div
        className="review"
      >
        <blockquote
          className="review__quote"
        >
          <p
            className="review__text"
          >
            1234
          </p>
          <footer
            className="review__details"
          >
            <cite
              className="review__author"
            >
              Isa
            </cite>
            <time
              className="review__date"
              dateTime="2016-12-24"
            >
              October 29, 2020
            </time>
          </footer>
        </blockquote>
        <div
          className="review__rating"
        >
          4.5
        </div>
      </div>
    </div>
  </div>
</div>
`;
